crypto/bio/bss_dgram.c

```diff
@@ -808,6 +992,627 @@ static int dgram_puts(BIO *bp, const char *str)
+        BIO_MSG_N(msg, i).flags    = 0;
+    }
+
+    return i > 0 ? (ossl_ssize_t)i : PACK_ERRNO(get_last_socket_error());
+
+#elif M_METHOD == M_METHOD_WSARECVMSG || M_METHOD == M_METHOD_RECVFROM
+# if M_METHOD == M_METHOD_WSARECVMSG
+    if (bio_WSASendMsg != NULL) {
+        /* WSASendMsg-based implementation for Windows. */
+        translate_msg_win(&wmsg, &wbuf, control, msg);
+
+        if (msg[0].local != NULL) {
+            if (!have_local_enabled)
+                return -1;
+
+            if (pack_local(b, &wmsg, msg[0].local) < 1)
+                return -1;
+        }
+
+        ret = WSASendMsg((SOCKET)b->num, &wmsg, 0, &num_bytes_sent, NULL, NULL);
+        if (ret < 0)
+            return PACK_ERRNO(get_last_socket_error());
+
+        msg[0].data_len = num_bytes_sent;
+        msg[0].flags    = 0;
+        return 1;
+    }
+# endif
+
+    /*
+     * Fallback to sendto and send a single message.
+     */
+    if (msg[0].local != NULL)
+        /*
+         * We cannot set the local address if using sendto
+         * so fail in this case
+         */
+        return -1;
+
+    ret = sendto(b->num, msg[0].data,
+# if defined(OPENSSL_SYS_WINDOWS)
+                 (int)msg[0].data_len,
+# else
+                 msg[0].data_len,
+# endif
+                 sysflags,
+                 msg[0].peer != NULL ? &msg[0].peer->sa : NULL,
+                 msg[0].peer != NULL ? sizeof(*msg[0].peer) : 0);
+    if (ret <= 0)
+        return PACK_ERRNO(get_last_socket_error());
+
+    msg[0].data_len = ret;
+    msg[0].flags    = 0;
+    return 1;
+
+#else
+    ERR_raise(ERR_LIB_BIO, BIO_R_UNSUPPORTED_METHOD);
+    return -2;
+#endif
+}
+
+static ossl_ssize_t dgram_recvmmsg(BIO *b, BIO_MSG *msg, size_t stride, size_t num_msg, uint64_t flags)
+{
+#if M_METHOD != M_METHOD_NONE
+    int ret;
+#endif
+#if M_METHOD == M_METHOD_RECVMMSG
+# define BIO_MAX_MSGS_PER_CALL   64
+    int sysflags;
+    bio_dgram_data *data = (bio_dgram_data *)b->ptr;
+    size_t i, j, jlim;
+    ossl_ssize_t num_done = 0;
+    struct mmsghdr mh[BIO_MAX_MSGS_PER_CALL];
+    struct iovec iov[BIO_MAX_MSGS_PER_CALL];
+    unsigned char control[BIO_CMSG_ALLOC_LEN][BIO_MAX_MSGS_PER_CALL];
+    int have_local_enabled = data->local_addr_enabled;
+#elif M_METHOD == M_METHOD_RECVMSG
+    int sysflags;
+    bio_dgram_data *data = (bio_dgram_data *)b->ptr;
+    size_t i;
+    struct msghdr mh;
+    struct iovec iov;
+    unsigned char control[BIO_CMSG_ALLOC_LEN];
+    int have_local_enabled = data->local_addr_enabled;
+#elif M_METHOD == M_METHOD_WSARECVMSG
+    bio_dgram_data *data = (bio_dgram_data *)b->ptr;
+    int have_local_enabled = data->local_addr_enabled;
+    WSAMSG wmsg;
+    WSABUF wbuf;
+    DWORD num_bytes_received = 0;
+    unsigned char control[BIO_CMSG_ALLOC_LEN];
+#endif
+#if M_METHOD == M_METHOD_RECVFROM || M_METHOD == M_METHOD_WSARECVMSG
+    int sysflags;
+    socklen_t slen;
+#endif
+
+    if (num_msg == 0)
+        return 0;
```

Comment:
Should there be an error on the stack in this case? Similar comment for other error returns below.

Reply:
Ah - ok. So 0 here is actually "successfully sent 0 messages"?

My comment about errors on the stack still applies to other error returns thought.